Obsidian.IO/ProtocolReader.cs

```diff
@@ -0,0 +1,299 @@
+        {
+            var v = Unsafe.ReadUnaligned<long>(ref GetBufferRef());
+            if (BitConverter.IsLittleEndian)
+                v = BinaryPrimitives.ReverseEndianness(v);
+
+            index += sizeof(long);
+            return Unsafe.As<long, double>(ref v);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public int ReadVarInt()
+        {
+            var numRead = 0;
+            var result = 0;
+            byte read;
+
+            do
+            {
+                read = ReadByte();
+                var value = read & 0b01111111;
+                result |= value << (7 * numRead);
+
+                numRead++;
+                if (numRead > 5)
+                {
+                    throw new InvalidOperationException("VarInt is too big");
+                }
+            } while ((read & 0b10000000) != 0);
+
+            return result;
+        }
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public long ReadVarLong()
+        {
+            var numRead = 0;
+            var result = 0L;
+            byte read;
+
+            do
+            {
+                read = ReadByte();
+                var value = read & 0b01111111;
+                result |= (long)value << (7 * numRead);
+
+                numRead++;
+                if (numRead > 10)
+                {
+                    throw new InvalidOperationException("VarLong is too big");
+                }
+            } while ((read & 0b10000000) != 0);
+
+            return result;
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public string ReadVarString() => ReadString(ReadVarInt());
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public string ReadUShortString() => ReadString(ReadUInt16());
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public string ReadString(int count)
+        {
+            if (count == 0) return string.Empty;
+
+            if (count <= 1024)
+            {
+                Span<byte> span = stackalloc byte[count];
+                ReadSpan(span);
+                return Encoding.UTF8.GetString(span);
+            }
+            
+            var pool = ArrayPool<byte>.Shared;
+            var arr = pool.Rent(count);
+
+            var arrSpan = arr.AsSpan(0, count);
+            ReadSpan(arrSpan);
+            
+            var str = Encoding.UTF8.GetString(arrSpan);
+            pool.Return(arr);
+            return str;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Guid ReadGuid()
+        {
+            var id = Unsafe.ReadUnaligned<Guid>(ref GetBufferRef());
+            index += 16;
+            return id;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void ReadInt64Array(Span<long> span)
+        {
+            for (var i = 0; i < span.Length; i++)
+                span[i] = ReadInt64();
```

Comment:
Isn't there some method for this? What comes to mind is `Span<byte> MemoryMarshal.AsBytes<long>(span)` and then `Span.CopyTo`, but there might be something even better.